sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -473,10 +473,47 @@ class Analyzer(
    */
   object ImplicitGenerate extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = plan transform {
-      case Project(Seq(Alias(g: Generator, _)), child) =>
-        Generate(g, join = false, outer = false, None, child)
+      case Project(Seq(Alias(g: Generator, name)), child) =>
+        Generate(g, join = false, outer = false, child, qualifier = None, name :: Nil, Nil)
+      case Project(Seq(MultiAlias(g: Generator, names)), child) =>
+        Generate(g, join = false, outer = false, child, qualifier = None, names, Nil)
     }
   }
+
+  object ResolveGenerate extends Rule[LogicalPlan] {
+    // Construct the output attributes for the generator,
+    // The output attribute names can be either specified or
+    // auto generated.
+    private def makeGeneratorOutput(
+        generator: Generator,
+        attributeNames: Seq[String],
+        qualifier: Option[String]): Array[Attribute] = {
+      val elementTypes = generator.elementTypes
+
+      val raw = if (attributeNames.size == elementTypes.size) {
```

Comment:
Could this instead be:
 - attributeName.size == 0 => auto generate
 - attributeNames.size == elementTypes.size => use names
 - otherwise throw error

Lets see what hive does.